src/shell/stage-in.c

```diff
@@ -0,0 +1,477 @@
+/************************************************************\
+ * Copyright 2022 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* stage-in.c - copy previously mapped files for job */
+
+#define FLUX_SHELL_PLUGIN_NAME "stage-in"
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <unistd.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <libgen.h>
+#include <jansson.h>
+#include <argz.h>
+#include <archive.h>
+#include <archive_entry.h>
+#include <flux/core.h>
+
+#include "ccan/base64/base64.h"
+#include "src/common/libcontent/content.h"
+#include "src/common/libutil/errno_safe.h"
+#include "src/common/libutil/monotime.h"
+#include "src/common/libutil/fileref.h"
+
+#include "builtins.h"
+#include "internal.h"
+#include "info.h"
+
+struct stage_in {
+    json_t *tags;
+    const char *pattern;
+    const char *destdir;
+    flux_t *h;
+    int count;
+    size_t total_size;
+    int direct;
+};
+
+json_t *parse_tags (const char *s, const char *default_value)
+{
+    char *argz = NULL;
+    size_t argz_len;
+    json_t *a;
+    json_t *o;
+    const char *entry;
+
+    if (!(a = json_array ()))
+        return NULL;
+    if (s) {
+        if (argz_create_sep (s, ',', &argz, &argz_len) != 0)
+            goto error;
+        entry = NULL;
+        while ((entry = argz_next (argz, argz_len, entry))) {
+            if (!(o = json_string (entry))
+                || json_array_append_new (a, o) < 0) {
+                json_decref (o);
+                goto error;
+            }
+        }
+    }
+    if (json_array_size (a) == 0 && default_value) {
+        if (!(o = json_string (default_value))
+            || json_array_append_new (a, o) < 0) {
+            json_decref (o);
+            goto error;
+        }
+    }
+    free (argz);
+    return a;
+error:
+    free (argz);
+    json_decref (a);
+    return NULL;
+}
+
+/* Decode the raw data field a fileref object, setting the result in 'data'
+ * and 'data_size'.  Caller must free.
+ */
+static int decode_data (const char *s, void **data, size_t *data_size)
+{
+    if (s) {
+        int len = strlen (s);
+        size_t bufsize = base64_decoded_length (len);
+        void *buf;
+        ssize_t n;
+
+        if (!(buf = calloc (1, bufsize)))
+            return -1;
+        if ((n = base64_decode (buf, bufsize, s, len)) < 0) {
+            free (buf);
+            errno = EINVAL;
+            return -1;
+        }
```

Comment:
nit, malloc instead of calloc, since you're going to overwrite the buffer immediately

(i guess this was done in cmd/builtin/filemap.c too)

Reply:
Regarding the code duplication, yeah, since the two implementations have different styles of error handling, it seemed like a high effort to refactor.  I could be convinced if a few more things need to be changed in both places :-)